blockprod/src/detail/utils.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use chainstate::{
+    chainstate_interface::ChainstateInterface, BlockIndex, GenBlockIndex, NonZeroPoolBalances,
+};
+use chainstate_types::{pos_randomness::PoSRandomness, EpochData};
+use common::{
+    chain::{
+        block::timestamp::{BlockTimestamp, BlockTimestampInternalType},
+        ChainConfig, PoolId,
+    },
+    primitives::{Amount, BlockHeight},
+};
+
+use crate::BlockProductionError;
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub enum PoSAccountingError {
+    #[error("Staker balance retrieval error: {0}")]
+    StakerBalanceRetrievalError(String),
+}
+
+pub fn get_pool_staker_balance(
+    chainstate: &dyn ChainstateInterface,
```

Comment:
Is there a reason for using dyn all over this file, instead of a generic? Anything that derefs to dyn ChainstateInterface also implements ChainstateInterface, meaning that a reference or a box should be passed to a generic with trait bount ChainstateInterface with no issues.